modules/index/src/main/java/org/apache/ignite/internal/index/IndexTaskManager.java

```diff
@@ -0,0 +1,207 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import org.apache.ignite.internal.catalog.CatalogManager;
+import org.apache.ignite.internal.catalog.CatalogService;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexStatus;
+import org.apache.ignite.internal.catalog.events.CatalogEvent;
+import org.apache.ignite.internal.catalog.events.CreateIndexEventParameters;
+import org.apache.ignite.internal.catalog.events.RemoveIndexEventParameters;
+import org.apache.ignite.internal.catalog.events.StoppingIndexEventParameters;
+import org.apache.ignite.internal.close.ManuallyCloseable;
+import org.apache.ignite.internal.event.EventListener;
+import org.apache.ignite.internal.event.EventParameters;
+import org.apache.ignite.internal.network.ClusterService;
+import org.apache.ignite.internal.placementdriver.PlacementDriver;
+import org.apache.ignite.internal.placementdriver.ReplicaMeta;
+import org.apache.ignite.internal.placementdriver.event.PrimaryReplicaEvent;
+import org.apache.ignite.internal.placementdriver.event.PrimaryReplicaEventParameters;
+import org.apache.ignite.internal.replicator.TablePartitionId;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+
+/**
+ * Component that reacts to certain Catalog changes and starts or stops corresponding {@link ChangeIndexStatusTask}s via the
+ * {@link ChangeIndexStatusTaskScheduler}.
+ *
+ * <p>Tasks will be started only if the local node is the primary replica for the {@code 0} partition of the table which indexes have been
+ * modified in the Catalog.
+ *
+ * <p>The following events are being monitored:
+ * <ul>
+ *     <li>{@link CatalogEvent#INDEX_CREATE} - when an index is created, a task will be started to move the index to the
+ *     {@link CatalogIndexStatus#BUILDING} state; </li>
+ *     <li>{@link CatalogEvent#INDEX_STOPPING} - when an index is dropped, a task will be started to remove the index from the Catalog;</li>
+ *     <li>{@link CatalogEvent#INDEX_REMOVED} - when an index is removed from the Catalog, all ongoing tasks for the index will be stopped;
+ *     </li>
+ *     <li>{@link PrimaryReplicaEvent#PRIMARY_REPLICA_ELECTED} - if the local node has become the primary replica for partition {@code 0} of
+ *     the table, it starts corresponding tasks depending on the current status of indices in the Catalog. If the local node stops being the
+ *     primary replica for the partition {@code 0}, it stops all tasks that belong to indices of the table that the partition belongs to.
+ *     </li>
+ * </ul>
+ *
+ * <p>On node recovery, tasks will be started on {@link PrimaryReplicaEvent#PRIMARY_REPLICA_ELECTED}, which will fire due to a change in
+ * the {@link ReplicaMeta#getLeaseholderId()} on node restart but after {@link ReplicaMeta#getExpirationTime()}.
+ */
+class IndexTaskManager implements ManuallyCloseable {
```

Comment:
I think the name `IndexChangeStatusController` is more suitable.
And for `IndexBuildingManager` the name is `IndexChangeStatusManager`.

Reply:
What is the difference between a Manager and a Controller in this case? These names look very similar to me

Reply:
For me, now postfix "Manager" is associated with a component that extends `org.apache.ignite.internal.manager.IgniteComponent`, but in our case this is not the case.

Reply:
Renamed